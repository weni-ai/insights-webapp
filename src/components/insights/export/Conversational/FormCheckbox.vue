<template>
  <ExportCheckboxs
    data-testid="conversational-form-checkbox"
    :modelFields="translatedModelFields"
    :selectedFields="translatedSelectedFields"
    :enabledModels="translatedEnabledModels"
    :modelFilters="modelFilters"
    :isLoading="isLoading"
    @model-toggle="handleModelToggle"
    @field-toggle="handleFieldToggle"
  />
</template>

<script setup lang="ts">
import ExportCheckboxs from '../ExportCheckboxs.vue';
import { useConversationalExport } from '@/store/modules/export/conversational/export';
import { storeToRefs } from 'pinia';
import { onMounted, ref, computed } from 'vue';

const conversationalExport = useConversationalExport();
const {
  updateModelFieldSelection,
  toggleModelEnabled,
  initializeDefaultFields,
} = conversationalExport;
const { model_fields, selected_fields, enabled_models, custom_widgets } =
  storeToRefs(conversationalExport);
const isLoading = ref(false);

const modelFilters = computed(() => []);

const getUniqueDisplayName = (widget: any): string => {
  const baseName = widget.name;
  const shortUuid = widget.uuid.slice(0, 8);

  const duplicates = custom_widgets.value.filter((w) => w.name === baseName);

  if (duplicates.length > 1) {
    return `${baseName} (${shortUuid})`;
  }

  return baseName;
};

const translatedSelectedFields = computed(() => {
  const fields = { ...selected_fields.value };

  if (custom_widgets.value.length > 0) {
    custom_widgets.value.forEach((widget) => {
      if (fields[widget.uuid]) {
        const fieldData = fields[widget.uuid];
        delete fields[widget.uuid];
        const uniqueName = getUniqueDisplayName(widget);
        fields[uniqueName] = fieldData;
      }
    });
  }

  return fields;
});

const translatedEnabledModels = computed(() => {
  return enabled_models.value.map((modelKey) => {
    const widget = custom_widgets.value.find((w) => w.uuid === modelKey);
    return widget ? getUniqueDisplayName(widget) : modelKey;
  });
});

const translatedModelFields = computed(() => {
  const fields = { ...model_fields.value };

  if (custom_widgets.value.length > 0) {
    custom_widgets.value.forEach((widget) => {
      if (fields[widget.uuid]) {
        const fieldData = fields[widget.uuid];
        delete fields[widget.uuid];
        const uniqueName = getUniqueDisplayName(widget);
        fields[uniqueName] = fieldData;
      }
    });
  }

  return fields;
});

const initializeFields = () => {
  isLoading.value = true;
  try {
    initializeDefaultFields();
  } catch (error) {
    console.error('Error initializing conversational fields', error);
  } finally {
    isLoading.value = false;
  }
};

const getModelKeyForStore = (displayName: string): string => {
  let widget = custom_widgets.value.find((w) => w.name === displayName);

  if (!widget) {
    const match = displayName.match(/^(.+) \(([a-f0-9-]{8})\)$/);
    if (match) {
      const [, baseName, shortUuid] = match;
      widget = custom_widgets.value.find(
        (w) => w.name === baseName && w.uuid.startsWith(shortUuid),
      );
    }
  }

  return widget ? widget.uuid : displayName;
};

const handleModelToggle = (modelName: string, enabled: boolean) => {
  const storeKey = getModelKeyForStore(modelName);
  toggleModelEnabled(storeKey, enabled);
};

const handleFieldToggle = (
  modelName: string,
  fieldName: string,
  selected: boolean,
) => {
  const storeKey = getModelKeyForStore(modelName);
  updateModelFieldSelection(storeKey, fieldName, selected);
};

onMounted(() => {
  initializeFields();
});
</script>
