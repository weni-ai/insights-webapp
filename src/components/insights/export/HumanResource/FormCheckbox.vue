<template>
  <ExportCheckboxs
    :modelFields="model_fields"
    :selectedFields="selected_fields"
    :enabledModels="enabled_models"
    :sectors="sectors"
    :queues="queues"
    :agents="agents"
    :tags="tags"
    :isLoading="isLoading"
    @model-toggle="handleModelToggle"
    @field-toggle="handleFieldToggle"
  />
</template>

<script setup lang="ts">
import ExportCheckboxs from '../ExportCheckboxs.vue';
import exportService from '@/services/api/resources/export/export';
import { useHumanResourceExport } from '@/store/modules/export/humanResource/export';
import { storeToRefs } from 'pinia';
import { onMounted, ref } from 'vue';

const humanResourceExport = useHumanResourceExport();
const { setModelFields, updateModelFieldSelection, toggleModelEnabled } =
  humanResourceExport;
const {
  model_fields,
  selected_fields,
  enabled_models,
  sectors,
  queues,
  agents,
  tags,
} = storeToRefs(humanResourceExport);
const isLoading = ref(false);

const fetchModelFields = async () => {
  isLoading.value = true;
  try {
    const modelFields = await exportService.getModelFields();
    const transformedFields: any = {};

    Object.entries(modelFields).forEach(([modelName, fields]) => {
      transformedFields[modelName] = {};
      Object.entries(fields as any).forEach(
        ([fieldName, fieldData]: [string, any]) => {
          transformedFields[modelName][fieldName] = {
            type: fieldData.type,
          };
        },
      );
    });

    setModelFields(transformedFields);
  } catch (error) {
    console.error(error);
  } finally {
    isLoading.value = false;
  }
};

const handleModelToggle = (modelName: string, enabled: boolean) => {
  toggleModelEnabled(modelName, enabled);
};

const handleFieldToggle = (
  modelName: string,
  fieldName: string,
  selected: boolean,
) => {
  updateModelFieldSelection(modelName, fieldName, selected);
};

onMounted(() => {
  fetchModelFields();
});
</script>
