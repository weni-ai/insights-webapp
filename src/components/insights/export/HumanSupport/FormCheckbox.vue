<template>
  <ExportCheckboxs
    data-testid="human-support-form-checkbox"
    :modelFields="model_fields"
    :selectedFields="selected_fields"
    :enabledModels="enabled_models"
    :modelFilters="modelFilters"
    :isLoading="isLoading"
    @model-toggle="handleModelToggle"
    @field-toggle="handleFieldToggle"
  />
</template>

<script setup lang="ts">
import ExportCheckboxs from '../ExportCheckboxs.vue';
import exportService from '@/services/api/resources/export/humanSupport/export';
import { useHumanSupportExport } from '@/store/modules/export/humanSupport/export';
import { storeToRefs } from 'pinia';
import { onMounted, ref, computed } from 'vue';

const humanSupportExport = useHumanSupportExport();
const { setModelFields, updateModelFieldSelection, toggleModelEnabled } =
  humanSupportExport;
const {
  model_fields,
  selected_fields,
  enabled_models,
  sectors,
  queues,
  agents,
  tags,
} = storeToRefs(humanSupportExport);
const isLoading = ref(false);

const modelFilters = computed(() => [
  {
    modelName: 'sectors',
    filterData: sectors.value,
  },
  {
    modelName: 'queues',
    filterData: queues.value,
  },
  {
    modelName: 'users',
    filterData: agents.value,
  },
  {
    modelName: 'sector_tags',
    filterData: tags.value,
  },
]);

const fetchModelFields = async () => {
  isLoading.value = true;
  try {
    const modelFields = await exportService.getModelFields();
    const transformedFields: any = {};

    Object.entries(modelFields).forEach(([modelName, fields]) => {
      transformedFields[modelName] = {};
      Object.entries(fields as any).forEach(
        ([fieldName, fieldData]: [string, any]) => {
          transformedFields[modelName][fieldName] = {
            type: fieldData.type,
          };
        },
      );
    });

    setModelFields(transformedFields);
  } catch (error) {
    console.error(error);
  } finally {
    isLoading.value = false;
  }
};

const handleModelToggle = (modelName: string, enabled: boolean) => {
  toggleModelEnabled(modelName, enabled);
};

const handleFieldToggle = (
  modelName: string,
  fieldName: string,
  selected: boolean,
) => {
  updateModelFieldSelection(modelName, fieldName, selected);
};

onMounted(() => {
  fetchModelFields();
});
</script>
