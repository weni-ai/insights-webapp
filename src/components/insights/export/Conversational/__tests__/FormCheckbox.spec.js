import { beforeEach, describe, expect, it, vi } from 'vitest';
import { config, mount } from '@vue/test-utils';
import { createI18n } from 'vue-i18n';

import FormCheckbox from '../FormCheckbox.vue';

const mockStore = {
  model_fields: { value: {} },
  selected_fields: { value: {} },
  enabled_models: { value: [] },
  custom_widgets: { value: [] },
  updateModelFieldSelection: vi.fn(),
  toggleModelEnabled: vi.fn(),
  initializeDefaultFields: vi.fn(),
};

vi.mock('@/store/modules/export/conversational/export', () => ({
  useConversationalExport: () => mockStore,
}));

vi.mock('pinia', async (importOriginal) => {
  const actual = await importOriginal();
  return {
    ...actual,
    storeToRefs: (store) => store,
  };
});

const i18n = createI18n({
  legacy: false,
  locale: 'en',
  messages: { en: {} },
});

config.global.plugins = [i18n];

describe('Conversational FormCheckbox', () => {
  let wrapper;

  const createWrapper = (storeOverrides = {}) => {
    Object.assign(mockStore, storeOverrides);

    return mount(FormCheckbox, {
      global: {
        stubs: {
          ExportCheckboxs: true,
        },
      },
    });
  };

  beforeEach(() => {
    vi.clearAllMocks();

    Object.assign(mockStore, {
      model_fields: { value: {} },
      selected_fields: { value: {} },
      enabled_models: { value: [] },
      custom_widgets: { value: [] },
    });

    wrapper = createWrapper();
  });

  describe('Component rendering', () => {
    it('should render ExportCheckboxs component', () => {
      const component = wrapper.find(
        '[data-testid="conversational-form-checkbox"]',
      );
      expect(component.exists()).toBe(true);
    });
  });

  describe('Lifecycle hooks', () => {
    it('should initialize fields on mounted', () => {
      expect(mockStore.initializeDefaultFields).toHaveBeenCalled();
    });

    it('should handle initialization errors', () => {
      const consoleError = vi.spyOn(console, 'error').mockImplementation();
      mockStore.initializeDefaultFields.mockImplementation(() => {
        throw new Error('Init error');
      });

      wrapper = createWrapper();

      expect(consoleError).toHaveBeenCalled();
      consoleError.mockRestore();
    });
  });

  describe('Widget translation', () => {
    it('should translate widget UUID to display name', () => {
      wrapper = createWrapper({
        custom_widgets: {
          value: [
            { uuid: '123-456', name: 'Widget A' },
            { uuid: '789-abc', name: 'Widget B' },
          ],
        },
        model_fields: {
          value: { '123-456': { field1: {} }, '789-abc': { field2: {} } },
        },
      });

      const translated = wrapper.vm.translatedModelFields;
      expect(translated['Widget A']).toBeDefined();
      expect(translated['Widget B']).toBeDefined();
      expect(translated['123-456']).toBeUndefined();
    });

    it('should handle duplicate widget names', () => {
      wrapper = createWrapper({
        custom_widgets: {
          value: [
            { uuid: '12345678-1234', name: 'Widget' },
            { uuid: '87654321-4321', name: 'Widget' },
          ],
        },
        model_fields: {
          value: {
            '12345678-1234': { field1: {} },
            '87654321-4321': { field2: {} },
          },
        },
      });

      const translated = wrapper.vm.translatedModelFields;
      expect(translated['Widget (12345678)']).toBeDefined();
      expect(translated['Widget (87654321)']).toBeDefined();
    });
  });

  describe('Event handlers', () => {
    it('should handle model toggle with display name', () => {
      wrapper = createWrapper({
        custom_widgets: {
          value: [{ uuid: '123-456', name: 'Widget A' }],
        },
      });

      wrapper.vm.handleModelToggle('Widget A', true);
      expect(mockStore.toggleModelEnabled).toHaveBeenCalledWith(
        '123-456',
        true,
      );
    });

    it('should handle field toggle with display name', () => {
      wrapper = createWrapper({
        custom_widgets: {
          value: [{ uuid: '123-456', name: 'Widget A' }],
        },
      });

      wrapper.vm.handleFieldToggle('Widget A', 'field1', true);
      expect(mockStore.updateModelFieldSelection).toHaveBeenCalledWith(
        '123-456',
        'field1',
        true,
      );
    });

    it('should handle model toggle with UUID format', () => {
      wrapper = createWrapper({
        custom_widgets: {
          value: [{ uuid: '12345678-1234', name: 'Widget' }],
        },
      });

      wrapper.vm.handleModelToggle('Widget (12345678)', true);
      expect(mockStore.toggleModelEnabled).toHaveBeenCalledWith(
        '12345678-1234',
        true,
      );
    });
  });

  describe('Computed properties', () => {
    it('should compute modelFilters as empty array', () => {
      expect(wrapper.vm.modelFilters).toEqual([]);
    });

    it('should translate enabled models', () => {
      wrapper = createWrapper({
        custom_widgets: {
          value: [{ uuid: '123-456', name: 'Widget A' }],
        },
        enabled_models: { value: ['123-456'] },
      });

      expect(wrapper.vm.translatedEnabledModels).toEqual(['Widget A']);
    });

    it('should translate selected fields', () => {
      wrapper = createWrapper({
        custom_widgets: {
          value: [{ uuid: '123-456', name: 'Widget A' }],
        },
        selected_fields: { value: { '123-456': ['field1'] } },
      });

      const translated = wrapper.vm.translatedSelectedFields;
      expect(translated['Widget A']).toEqual(['field1']);
      expect(translated['123-456']).toBeUndefined();
    });
  });

  describe('Component structure', () => {
    it('should match snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
