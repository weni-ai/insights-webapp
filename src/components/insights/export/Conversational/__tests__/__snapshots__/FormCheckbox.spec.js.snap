// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Conversational FormCheckbox > Component structure > should match snapshot 1`] = `
<export-checkboxs-stub
  data-testid="conversational-form-checkbox"
  enabledmodels=""
  isloading="false"
  modelfields="[object Object]"
  modelfilters=""
  selectedfields="[object Object]"
/>
`;
